File: java/std_nat/src/fan/std/Lazy.java
package fan.std;

import fan.sys.FanObj;
import fan.sys.Func;
import fanx.main.Sys;
import fanx.main.Type;

public class Lazy extends FanObj {
	private Object val;
	private Func initial;
	
	private static Type type = null;

	public Type typeof() {
		if (type == null) {
			type = Sys.findType("std::Lazy");
		}
		return type;
	}
	
	public static Lazy make(Func init) {
		Lazy self = new Lazy();
		self.initial = (Func)init.toImmutable();
		return self;
	}
	
	public Object get() {
		if (val == null) {
			synchronized(this) {
				if (val == null) {
					Object v = initial.call();
					val = FanObj.toImmutable(v);
				}
			}
		}
		return val;
	}
}

File: src/std/js/Lazy.js



fan.std.Lazy = fan.sys.Obj.$extend(fan.sys.Obj);
fan.std.Lazy.prototype.$ctor = function() {}
fan.std.Lazy.prototype.$typeof = function() { return fan.std.Lazy.$type; }

fan.std.Lazy.make = function(init) {
	var self = new fan.std.Lazy();
	self.initial = init.toImmutable();
	return self;
}

fan.std.Lazy.prototype.get = function() {
	if (this.m_val == null) {
		v = initial.call();
		this.m_val = fan.sys.ObjUtil.toImmutable(v);
	}
	return this.m_val;
}

